napari/experimental/_progressive_loading.py

```diff
@@ -0,0 +1,1652 @@
+                    f"render_sequence: yielding chunk {chunk_slice} at scale {scale} which has priority\t{priority}"
+                )
+
+                yield tuple(list(chunk_result) + [len(chunk_queue) == 0])
+
+                # TODO blank out lower resolution
+                # if lower resolution is visible, send zeros
+
+            LOGGER.info(f"render_sequence: done fetching {scale}")
+
+
+def chunk_keys_within_interval(chunk_keys, mins, maxs):
+    """Return chunk_keys that are within interval.
+
+    Returns a dictionary with a list of slices for each dimension
+    """
+    # contained_keys is an array with list of slices contained along each
+    # dimension
+    contained_keys: Dict = defaultdict(list)
+    for dim, chunk_slices in enumerate(chunk_keys):
+        for sl in chunk_slices:
+            below_min = sl.start < mins[dim]
+            above_max = sl.stop > maxs[dim]
+            # If start and stop are below interval, or
+            #    start and stop are above interval: return False
+            if (below_min and sl.stop < mins[dim]) or (
+                above_max and sl.start > maxs[dim]
+            ):
+                return []
+            else:
+                contained_keys[dim] += [sl]
+
+    return contained_keys
+
+
+def get_layer_name_for_scale(scale):
+    """Return the layer name for a given scale."""
+    return f"scale_{scale}"
+
+
+@tz.curry
+def dims_update_handler(invar, data=None, viewer=None, ndisplay=None):
+    """Start a new render sequence with the current viewer state.
+
+    Parameters
+    ----------
+    invar : Event or viewer
+        either an event or a viewer
+    full_shape : tuple
+        a tuple representing the shape of the highest resolution array
+    """
+    # The root layer corresponds to the highest resolution
+    root_layer = viewer.layers[get_layer_name_for_scale(0)]
+
+    worker = None
+    if "worker" in root_layer.metadata:
+        worker = root_layer.metadata["worker"]
+
+    # TODO global worker usage is not viable for real implementation
+    # Terminate existing multiscale render pass
+    if worker:
+        # TODO this might not terminate threads properly
+        worker.await_workers()
+        # worker.await_workers(msecs=30000)
+
+    # Find the corners of visible data in the highest resolution
+    corner_pixels = root_layer.corner_pixels
+
+    top_left = np.max((corner_pixels,), axis=0)[0, :]
+    bottom_right = np.min((corner_pixels,), axis=0)[1, :]
+
+    camera = viewer.camera.copy()
+
+    # TODO Added to skip situations when 3D isnt setup on layer yet??
+    if np.any((bottom_right - top_left) == 0):
+        return
+
+    # TODO we could add padding around top_left and bottom_right to account
+    #      for future camera movement
+
+    # Interval must be nonnegative
+    if not np.all(top_left <= bottom_right):
+        import pdb
+
+        pdb.set_trace()
+
+    corners = viewer.layers[get_layer_name_for_scale(0)].corner_pixels
```

Comment:
This is on the main thread, so we can assume layers weren't added/removed right?

```suggestion
    corners = root_layer.corner_pixels
```

Reply:
Actually, wait isn't `corners` just equivalent to `corner_pixels` above. If so, we could just delete this line and use `corner_pixels` instead? Maybe I'm missing something here...